File: shoring/fxapp/src/main/java/com/panopset/fxapp/SceneUpdater.java
package com.panopset.fxapp;

import com.panopset.compat.Logop;

/**
 * 
 * Call triggerAnUpdate after user input has changed,
 * override doUpdate to do the update.
 *
 */
public abstract class SceneUpdater {

  protected abstract void doUpdate();

  protected void triggerAnUpdate() {
    needsUpdate = true;
    if (updater == null) {
      createUpdater();
    }
  }
  
  private boolean needsUpdate = false;
  private boolean updating = false;
  private Thread updater;

  private synchronized void synchronizedUpdate() {
    needsUpdate = false;
    updating = true;
    doUpdate();
    updating = false;
  }

  private void submitUpdate() {
    if (!updating) {
      synchronizedUpdate();
    }
  }
  
  private synchronized void createUpdater() {
    if (updater != null) {
      return;
    }
    updater = new Thread(() -> {
      while (JavaFXapp.isActive()) {
        try {
          if (needsUpdate) {
            submitUpdate();
          }
          Thread.sleep(getSleepMillis());
        } catch (InterruptedException e) {
          Logop.warn(e);
          Thread.currentThread().interrupt();
        }
      }
    });
    updater.start();
  }

  private static final int DEFAULT_SLEEP_MILLIS = 1000;
  
  private Integer sleepMillis;
  
  public int getSleepMillis() {
    if (sleepMillis == null) {
      sleepMillis = DEFAULT_SLEEP_MILLIS;
    }
    return sleepMillis;
  }

  public void setSleepMillis(int value) {
    sleepMillis = value;
  }
}

File: code/shoring/fxapp/src/main/kotlin/com/panopset/fxapp/SceneUpdater.kt
package com.panopset.fxapp

import com.panopset.compat.Logz
import com.panopset.compat.Zombie

/**
 *
 * Call triggerAnUpdate after user input has changed,
 * override doUpdate to do the update.
 *
 */
abstract class SceneUpdater(val fxDoc: FxDoc) {
    protected abstract fun doUpdate()
    protected fun triggerAnUpdate() {
        needsUpdate = true
        if (updater == null) {
            createUpdater()
        }
    }

    private var needsUpdate = false
    private var updating = false
    private var updater: Thread? = null
    @Synchronized
    private fun synchronizedUpdate() {
        needsUpdate = false
        updating = true
        doUpdate()
        updating = false
    }

    private fun submitUpdate() {
        if (!updating) {
            synchronizedUpdate()
        }
    }

    @Synchronized
    private fun createUpdater() {
        if (updater != null) {
            return
        }
        updater = Thread {
            while (Zombie.isActive) {
                try {
                    if (needsUpdate) {
                        submitUpdate()
                    }
                    Thread.sleep(1000)
                } catch (e: InterruptedException) {
                    Logz.errorEx(e)
                    Thread.currentThread().interrupt()
                }
            }
        }
        updater!!.start()
    }
}
